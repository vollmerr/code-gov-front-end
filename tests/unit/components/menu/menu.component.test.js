import React from 'react'
import { shallow } from 'enzyme'
import cloneDeep from 'lodash.clonedeep'

import { eventMap as wEventMap } from 'mocks/window'
import { eventMap as dEventMap } from 'mocks/document'
import Menu from 'components/menu/menu.component'

const props = {
  logoLight: 'logo-src-light',
  menu: [
    { name: 'menu-opt-1', expanded: false, links: [{ id: 'menu-link-1' }] },
    { name: 'menu-opt-2', expanded: true , links: [{ id: 'menu-link-2' }]},
  ],
  officialBanner: 'official-banner',
  siteTitle: 'test-title',
  toggleSearchDropdown: jest.fn(),
}

let wrapper
let instance
describe('components - Menu', () => {
  beforeEach(() => {
    // props gets mutated, make deep copy
    wrapper = shallow(<Menu {...cloneDeep(props)} />)
    instance = wrapper.instance()
  })

  describe('componentDidMount', () => {
    it('should attach a `scroll` event listener that sets `notAtTop` when not at the top', () => {
      document.body.scrollTop = 100
      wEventMap.scroll()
      expect(wrapper.state('notAtTop')).toBeTruthy()
    })

    it('should attach a `click` event listener that collapses the header', () => {
      instance.collapse = jest.fn()
      // mock a ref, shallow render doesnt get refs :(
      instance.header = { current: { contains: jest.fn() } }

      // click outside header - collapses
      dEventMap.click({ target: 'div' })
      expect(instance.collapse).toBeCalled()

      // click inside header - not collapse
      jest.resetAllMocks()
      dEventMap.click({ target: instance.header.current })
      expect(instance.collapse).not.toBeCalled()
    })
  })

  describe('onClickMenuOption', () => {
    it('should inverse (mutation) the passed option`s `expanded` property', () => {
      const option = { expanded: true, links: [{ id: 'link' }] }
      // true => false
      instance.onClickMenuOption(option)
      expect(option.expanded).toBeFalsy()
      // false => true
      instance.onClickMenuOption(option)
      expect(option.expanded).toBeTruthy()
    })

    it('should set the `height` based off the number of links if expanded', () => {
      const option = { expanded: false, links: [{ id: 'link-1' }, { id: 'link-2' }] }
      instance.onClickMenuOption(option)
      expect(wrapper.state('height')).toBe(74 + 40 * 2)
    })

    it('should set the `height` as `auto` if not expanded', () => {
      const option = { expanded: true, links: [{ id: 'link-1' }] }
      instance.onClickMenuOption(option)
      expect(wrapper.state('height')).toBe('auto')
    })

    xit('should set `expanded` based off if any options are expanded', () => {
      // BUG: not set based off current option being expanded, based off outdated state
      // Remove (all references to expanded in state)? - doesnt look like even needed / adding `expanded` class to nav does nothing
    })
  })

  describe('collapse', () => {
    it('should set the menu as not being expanded', () => {
      wrapper.setState({ expanded: true })
      instance.collapse()
      expect(wrapper.state('expanded')).toBeFalsy()
    })

    it('should set the menu`s options as not being expanded', () => {
      wrapper.setState({ menu: [{ expanded: true }, { expanded: true }] })
      instance.collapse()
      expect(wrapper.state('menu').every(x => !x.expanded)).toBeTruthy()
    })

    it('should change the `height` to `auto`', () => {
      wrapper.setState({ height: 100 })
      instance.collapse()
      expect(wrapper.state('height')).toBe('auto')
    })
  })

  describe('render', () => {
    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })
})