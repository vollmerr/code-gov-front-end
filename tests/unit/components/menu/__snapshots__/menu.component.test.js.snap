// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components - Menu render should render correctly 1`] = `
<header
  className="main menu-background"
>
  <nav
    aria-label="primary"
    className="main menu-background"
    style={
      Object {
        "height": "auto",
      }
    }
  >
    <Connect(MobileMenuControl) />
    <OfficialBanner
      bannerOptions="official-banner"
    />
    <div
      className="indented"
    >
      <Link
        className="svg-container"
        replace={false}
        title="test-title Home"
        to="/home"
      >
        <img
          alt="code.ca.gov"
          className="home-logo"
          src="logo-src-light"
        />
      </Link>
    </div>
    <ul
      aria-label="primary"
      className="nav-menu"
      role="menubar"
    >
      <li
        className=""
        key="menu-opt-1"
        role="none"
      >
        <PrimaryMenuOption
          menuOption={
            Object {
              "expanded": false,
              "links": Array [
                Object {
                  "id": "menu-link-1",
                },
              ],
              "name": "menu-opt-1",
            }
          }
          onClick={[Function]}
        />
        <SecondaryDropdown
          menuOption={
            Object {
              "expanded": false,
              "links": Array [
                Object {
                  "id": "menu-link-1",
                },
              ],
              "name": "menu-opt-1",
            }
          }
          onClick={[Function]}
        />
      </li>
      <li
        className="expanded"
        key="menu-opt-2"
        role="none"
      >
        <PrimaryMenuOption
          menuOption={
            Object {
              "expanded": true,
              "links": Array [
                Object {
                  "id": "menu-link-2",
                },
              ],
              "name": "menu-opt-2",
            }
          }
          onClick={[Function]}
        />
        <SecondaryDropdown
          menuOption={
            Object {
              "expanded": true,
              "links": Array [
                Object {
                  "id": "menu-link-2",
                },
              ],
              "name": "menu-opt-2",
            }
          }
          onClick={[Function]}
        />
      </li>
    </ul>
  </nav>
</header>
`;
